File: src/travdata/cli/cmds/extractcsvtables.py
# -*- coding: utf-8 -*-
"""
Extracts data tables from the Mongoose Traveller 2022 core rules PDF as
CSV files.
"""

import argparse
import contextlib
import pathlib
import sys
import textwrap
from typing import Callable, Iterator

from progress import bar as progress  # type: ignore[import-untyped]
from travdata import config, filesio
from travdata.cli import cliutil
from travdata.extraction import bookextract
from travdata.extraction.pdf import cachingreader, tabulareader


def add_subparser(subparsers) -> None:
    """Adds a subcommand parser to ``subparsers``."""
    argparser: argparse.ArgumentParser = subparsers.add_parser(
        "extractcsvtables",
        description=__doc__,
        formatter_class=argparse.RawTextHelpFormatter,
        prefix_chars="-+",
    )
    argparser.set_defaults(run=run)

    argparser.add_argument(
        "book_name",
        help=textwrap.dedent(
            """
            Name identifier of the PDF file to extract.

            Use `travdata_cli -c CONFIG_DIR listbooks` to list accepted values
            for this argument.
            """
        ),
        metavar="BOOK",
    )
    argparser.add_argument(
        "input_pdf",
        help="Path to the PDF file to read tables from.",
        type=pathlib.Path,
        metavar="INPUT.PDF",
    )
    argparser.add_argument(
        "output",
        help=textwrap.dedent(
            """
            Path to the directory or ZIP file to output the CSV files into.

            Whether this is a directory or ZIP file is controlled by
            --output-type.
            """
        ),
        type=pathlib.Path,
        metavar="OUTPUT_PATH",
    )

    config.add_config_flag(argparser)

    argparser.add_argument(
        "--no-progress",
        help="""Disable progress bar.""",
        action="store_true",
        default=False,
    )
    argparser.add_argument(
        "--no-table-cache",
        help="""Disable the table cache.""",
        action="store_true",
        default=False,
    )

    argparser.add_argument(
        "--output-type",
        help=textwrap.dedent(
            """
            Controls how data is output to the OUTPUT_PATH.

            * AUTO guesses, based on any existing file or directory at the path
              or the path suffix ending in ".zip".
            * DIR writes as a directory.
            * ZIP writes as a ZIP file.
            """
        ),
        type=filesio.IOType,
        choices=filesio.IOType,
        default=filesio.IOType.AUTO,
    )

    outsel_grp = argparser.add_argument_group(
        "Output selection",
        description="Controls which data is extracted from the book.",
    )
    outsel_grp.add_argument(
        "--overwrite-existing",
        help=textwrap.dedent(
            """
            Extract CSV tables that already exist in the output. This is useful
            when testing larger scale changes to the configuration or code.
            """
        ),
        action="store_true",
        default=False,
    )
    outsel_grp.add_argument(
        "+t",
        "--with-tag",
        dest="with_tag",
        nargs="*",
        metavar="TAG",
        default=[],
        help=textwrap.dedent(
            """
            Only extract tables that have any of these tags. --without-tag takes
            precedence over this.
            """
        ),
    )
    outsel_grp.add_argument(
        "-t",
        "--without-tag",
        dest="without_tag",
        nargs="*",
        metavar="TAG",
        default=[],
        help=textwrap.dedent(
            """
            Only extract tables that do not have any of these tags. This takes
            precedence over --with-tag.
            """
        ),
    )

    tab_grp = argparser.add_argument_group("Tabula")
    tab_grp.add_argument(
        "--tabula-force-subprocess",
        help=textwrap.dedent(
            """
            If jpype cannot use libjvm, try seting this flag to use a slower
            path that uses Java as a subprocess.
            """
        ),
        action="store_true",
        default=False,
    )


@contextlib.contextmanager
def _progress_reporter(no_progress: bool) -> Iterator[Callable[[bookextract.ProgressEvent], None]]:
    if no_progress:
        progress_bar = None

        def on_progress(p: bookextract.ProgressEvent) -> None:
            del p  # unused

    else:
        progress_bar = progress.Bar("Extracting tables")
        progress_bar.start()

        def on_progress(p: bookextract.ProgressEvent) -> None:
            progress_bar.index = p.completed
            progress_bar.max = p.total
            progress_bar.update()

    try:
        yield on_progress
    finally:
        if progress_bar is not None:
            progress_bar.finish()


def _output_io_type_path(
    args: argparse.Namespace,
) -> filesio.IOTypePath:
    return filesio.IOTypePath(
        io_type=args.output_type,
        path=args.output,
    ).resolve_auto()


def run(args: argparse.Namespace) -> int:
    """CLI entry point."""

    with_tags = frozenset(args.with_tag)
    without_tags = frozenset(args.without_tag)
    if intersection := with_tags & without_tags:
        fmt_inter = ", ".join(sorted(intersection))
        print(
            f"Tags have been specified for both inclusion and exclusion: {fmt_inter}.",
            file=sys.stderr,
        )
        return 1

    ext_cfg = bookextract.ExtractionConfig(
        cfg_reader_type_path=config.config_reader_type_path(args),
        out_writer_type_path=_output_io_type_path(args),
        input_pdf=args.input_pdf,
        book_id=args.book_name,
        overwrite_existing=args.overwrite_existing,
        with_tags=with_tags,
        without_tags=without_tags,
    )

    with (
        tabulareader.TabulaClient(force_subprocess=args.tabula_force_subprocess) as tabula_client,
        cachingreader.optional_table_cache(
            delegate=tabula_client,
            disable=args.no_table_cache,
        ) as table_reader,
        _progress_reporter(args.no_progress) as on_progress,
    ):
        events = bookextract.extract_book(
            table_reader=table_reader,
            ext_cfg=ext_cfg,
            do_continue=lambda: True,
        )
        for event in events:
            match event:
                case bookextract.ProgressEvent() as progress_event:
                    on_progress(progress_event)
                case bookextract.ErrorEvent(message=message):
                    print(message, file=sys.stderr)
                case bookextract.EndedEvent(abnormal=abnormal):
                    if abnormal:
                        return cliutil.EX_SOFTWARE
                case _:
                    pass

    return 0

File: rust/src/cli/extractcsvtables.rs
use std::{
    path::{Path, PathBuf},
    sync::{atomic::AtomicBool, Arc},
};

use anyhow::{Context, Result};
use clap::Args;
use simple_bar::ProgressBar;

use crate::{
    extraction::{
        bookextract::{ExtractEvents, ExtractSpec, Extractor},
        pdf::TableReaderArgs,
    },
    filesio,
};

/// Extracts data tables from the Mongoose Traveller 2022 core rules PDF as CSV
/// files.
#[derive(Args, Debug)]
pub struct Command {
    /// Path to the configuration.
    book_name: String,

    /// Path to input PDF.
    input_pdf: PathBuf,

    /// Path to the directory or ZIP file to output the CSV files into.
    ///
    /// Whether this is a directory or ZIP file is controlled by --output-type.
    output: PathBuf,

    /// Path to the configuration. This must be either a directory or ZIP file,
    /// directly containing a config.yaml file, book.yaml files in directories,
    /// and its required Tabula templates. Some configurations for this should
    /// be included with this program's distribution.
    #[arg(long)]
    config: PathBuf,

    /// Controls how data is written to the output.
    ///
    /// By default, it guesses, based on any existing file or directory at the
    /// path or the path suffix ending in ".zip".
    #[arg(long)]
    output_type: Option<crate::filesio::IoType>,

    /// Extract CSV tables that already exist in the output. This is useful when
    /// testing larger scale changes to the configuration or code.
    #[arg(long)]
    overwrite_existing: bool,

    /// Only extract tables that have any of these tags. --without-tag takes
    /// precedence over this.
    #[arg(long, value_delimiter(','))]
    with_tags: Vec<String>,

    /// Only extract tables that do not have any of these tags. This takes
    /// precedence over --with-tag.
    #[arg(long, value_delimiter(','))]
    without_tags: Vec<String>,

    /// Show a progress bar reflecting overall extraction progress.
    #[arg(long, default_value = "true")]
    show_progress: bool,

    /// Options relating to configuring the table reader.
    #[command(flatten)]
    table_reader: TableReaderArgs,
}

/// Runs the subcommand.
pub fn run(cmd: &Command, xdg_dirs: xdg::BaseDirectories) -> Result<()> {
    let table_reader = cmd.table_reader.build(&xdg_dirs)?;

    let cfg_type = filesio::IoType::resolve_auto(None, &cmd.config);
    let cfg_reader = cfg_type
        .new_reader(&cmd.config)
        .with_context(|| format!("opening config path {:?} as {:?}", cmd.config, cfg_type))?;

    let output_type = filesio::IoType::resolve_auto(cmd.output_type, &cmd.output);
    let out_writer = output_type
        .new_read_writer(&cmd.output)
        .with_context(|| format!("opening output path {:?} as {:?}", cmd.output, output_type))?;

    let mut extractor = Extractor::new(table_reader.as_ref(), cfg_reader, out_writer)?;

    let spec = ExtractSpec {
        book_name: &cmd.book_name,
        input_pdf: &cmd.input_pdf,
        overwrite_existing: cmd.overwrite_existing,
        with_tags: &cmd.with_tags,
        without_tags: &cmd.without_tags,
    };

    let continue_intent = Arc::new(AtomicBool::new(true));
    let mut events = EventDisplayer::new(cmd.show_progress, continue_intent.clone())?;
    ctrlc::set_handler(move || continue_intent.store(false, std::sync::atomic::Ordering::SeqCst))?;

    extractor.extract_book(spec, &mut events);

    extractor.close()?;

    if let Err(err) = table_reader.close() {
        log::warn!("Failed to shut down table reader: {err}");
    }

    Ok(())
}

struct EventDisplayer {
    show_progress: bool,
    progress_bar: Option<ProgressBar>,
    continue_intent: Arc<AtomicBool>,
}

impl EventDisplayer {
    fn new(show_progress: bool, continue_intent: Arc<AtomicBool>) -> Result<Self> {
        Ok(EventDisplayer {
            show_progress,
            progress_bar: None,
            continue_intent: continue_intent.clone(),
        })
    }
}

impl ExtractEvents for EventDisplayer {
    fn on_progress(&mut self, _path: &Path, _completed: usize, total: usize) {
        if !self.show_progress {
            return;
        }

        let progress_bar: &mut ProgressBar = match self.progress_bar.as_mut() {
            Some(progress_bar) => progress_bar,
            None => {
                let progress_bar = ProgressBar::cargo_style(total as u32, 80, true);
                self.progress_bar = Some(progress_bar);
                self.progress_bar.as_mut().unwrap()
            }
        };
        progress_bar.update();
    }

    fn on_error(&mut self, err: anyhow::Error) {
        eprintln!("Error during extraction: {:?}.", err);
    }

    fn on_end(&mut self) {}

    fn do_continue(&self) -> bool {
        self.continue_intent
            .load(std::sync::atomic::Ordering::Relaxed)
    }
}
